From f1ee3ba1895ddfb5e9250a4f5f578d3a81cd3647 Mon Sep 17 00:00:00 2001
From: Sandor Yu <R01008@freescale.com>
Date: Wed, 18 Nov 2015 11:43:59 +0800
Subject: [PATCH 0674/5242] MLK-11832: MIPI CSI: Adjust mipi csi sensor power
 up process

commit  7d1d267d5710759824fc6103d60f8d0e68460069 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add s_stream interface for MIPI CSI and ov5647/ov5640 drivers.
Enable MIPI CSI sensor after MIPI CSI DPHY CLK enable
in case MIPI DPHY miss MIPI CSI sensor signal.

Signed-off-by: Sandor Yu <R01008@freescale.com>
(cherry picked from commit 26d81b416be6e3af203d1254a3116dbcf39a5605)
Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/mxc/capture/mxc_mipi_csi.c  |    2 ++
 .../media/platform/mxc/capture/ov5640_mipi_v2.c    |   15 ++++++++++-----
 drivers/media/platform/mxc/capture/ov5647_mipi.c   |   14 ++++++++++----
 3 files changed, 22 insertions(+), 9 deletions(-)

diff --git a/drivers/media/platform/mxc/capture/mxc_mipi_csi.c b/drivers/media/platform/mxc/capture/mxc_mipi_csi.c
index 62b0773..1c71de5 100644
--- a/drivers/media/platform/mxc/capture/mxc_mipi_csi.c
+++ b/drivers/media/platform/mxc/capture/mxc_mipi_csi.c
@@ -663,8 +663,10 @@ static int mipi_csis_s_stream(struct v4l2_subdev *mipi_sd, int enable)
 			goto unlock;
 		}
 		mipi_csis_start_stream(state);
+		v4l2_subdev_call(state->sensor_sd, video, s_stream, true);
 		state->flags |= ST_STREAMING;
 	} else {
+		v4l2_subdev_call(state->sensor_sd, video, s_stream, false);
 		mipi_csis_stop_stream(state);
 		state->flags &= ~ST_STREAMING;
 		if (debug > 0)
diff --git a/drivers/media/platform/mxc/capture/ov5640_mipi_v2.c b/drivers/media/platform/mxc/capture/ov5640_mipi_v2.c
index ca3bc52..62af8e5 100644
--- a/drivers/media/platform/mxc/capture/ov5640_mipi_v2.c
+++ b/drivers/media/platform/mxc/capture/ov5640_mipi_v2.c
@@ -871,7 +871,6 @@ static void OV5640_stream_off(void)
 	ov5640_write_reg(0x4202, 0x0f);
 }
 
-
 static int OV5640_get_sysclk(void)
 {
 	 /* calculate sysclk */
@@ -1304,8 +1303,6 @@ static int ov5640_change_mode_exposure_calc(enum ov5640_frame_rate frame_rate,
 	}
 	OV5640_set_shutter(cap_shutter);
 
-	OV5640_stream_on();
-
 err:
 	return retval;
 }
@@ -1345,8 +1342,6 @@ static int ov5640_change_mode_direct(enum ov5640_frame_rate frame_rate,
 	if (retval < 0)
 		goto err;
 
-	OV5640_stream_on();
-
 	OV5640_turn_on_AE_AG(1);
 
 err:
@@ -1756,9 +1751,19 @@ static int init_device(void)
 	return ret;
 }
 
+static int ov5640_s_stream(struct v4l2_subdev *sd, int enable)
+{
+	if (enable)
+		OV5640_stream_on();
+	else
+		OV5640_stream_off();
+	return 0;
+}
+
 static struct v4l2_subdev_video_ops ov5640_subdev_video_ops = {
 	.g_parm = ov5640_g_parm,
 	.s_parm = ov5640_s_parm,
+	.s_stream = ov5640_s_stream,
 };
 
 static const struct v4l2_subdev_pad_ops ov5640_subdev_pad_ops = {
diff --git a/drivers/media/platform/mxc/capture/ov5647_mipi.c b/drivers/media/platform/mxc/capture/ov5647_mipi.c
index 453478f..258213a 100644
--- a/drivers/media/platform/mxc/capture/ov5647_mipi.c
+++ b/drivers/media/platform/mxc/capture/ov5647_mipi.c
@@ -1047,8 +1047,6 @@ static int ov5647_change_mode_exposure_calc(enum ov5647_frame_rate frame_rate,
 	}
 	ov5647_set_shutter(cap_shutter);
 
-	ov5647_stream_on();
-
 err:
 	return retval;
 }
@@ -1088,8 +1086,6 @@ static int ov5647_change_mode_direct(enum ov5647_frame_rate frame_rate,
 	if (retval < 0)
 		goto err;
 
-	ov5647_stream_on();
-
 	ov5647_turn_on_AE_AG(1);
 
 err:
@@ -1501,9 +1497,19 @@ static int init_device(void)
 	return ret;
 }
 
+static int ov5647_s_stream(struct v4l2_subdev *sd, int enable)
+{
+	if (enable)
+		ov5647_stream_on();
+	else
+		ov5647_stream_off();
+	return 0;
+}
+
 static struct v4l2_subdev_video_ops ov5647_subdev_video_ops = {
 	.g_parm = ov5647_g_parm,
 	.s_parm = ov5647_s_parm,
+	.s_stream = ov5647_s_stream,
 };
 
 static const struct v4l2_subdev_pad_ops ov5647_subdev_pad_ops = {
-- 
1.7.9.5

